File: gtars-vrs/examples/gnomad_vrs.py
#!/usr/bin/env python3
"""Compute VRS Allele IDs from a gnomAD VCF using a real reference genome.

Usage:
    python gnomad_vrs.py <fasta_path> <vcf_path>
"""

import sys
import time

from gtars.refget import RefgetStore


def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <fasta_path> <vcf_path>", file=sys.stderr)
        sys.exit(1)

    fasta_path = sys.argv[1]
    vcf_path = sys.argv[2]

    # Step 1: Load reference genome
    print(f"Loading FASTA: {fasta_path}", file=sys.stderr)
    t0 = time.time()
    store = RefgetStore.in_memory()
    store.add_sequence_collection_from_fasta(fasta_path)
    load_time = time.time() - t0
    print(f"Loaded in {load_time:.1f}s", file=sys.stderr)

    # Get collection digest
    collections = store.list_collections()
    assert len(collections) > 0, "No collections found"
    collection_digest = collections[0].digest
    print(
        f"Collection: {collection_digest} ({collections[0].n_sequences} sequences)",
        file=sys.stderr,
    )

    # Step 2: Compute VRS IDs
    print(f"Processing VCF: {vcf_path}", file=sys.stderr)
    t1 = time.time()
    results = store.compute_vrs_ids(collection_digest, vcf_path)
    vrs_time = time.time() - t1

    # Step 3: Print results
    print("chrom\tpos\tref\talt\tvrs_id")
    for r in results:
        print(f"{r['chrom']}\t{r['pos']}\t{r['ref']}\t{r['alt']}\t{r['vrs_id']}")

    total_time = time.time() - t0
    print(
        f"\nComputed {len(results)} VRS IDs in {vrs_time:.3f}s "
        f"({len(results) / vrs_time:.0f} variants/sec)",
        file=sys.stderr,
    )
    print(f"Total time: {total_time:.1f}s", file=sys.stderr)


if __name__ == "__main__":
    main()

File: gtars-vrs/examples/gnomad_vrs.rs
//! Compute VRS Allele IDs from a gnomAD VCF using a real reference genome.
//!
//! Usage:
//!   cargo run --release --example gnomad_vrs -- <fasta_path> <vcf_path>

use std::time::Instant;

use gtars_refget::store::{FastaImportOptions, RefgetStore};
use gtars_vrs::vcf::compute_vrs_ids_streaming;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} <fasta_path> <vcf_path>", args[0]);
        std::process::exit(1);
    }
    let fasta_path = &args[1];
    let vcf_path = &args[2];

    // Step 1: Load reference genome into RefgetStore
    eprintln!("Loading FASTA: {}", fasta_path);
    let t0 = Instant::now();
    let mut store = RefgetStore::in_memory();
    store
        .add_sequence_collection_from_fasta(fasta_path, FastaImportOptions::new())
        .expect("Failed to load FASTA");
    let load_time = t0.elapsed();
    eprintln!("Loaded in {:.1}s", load_time.as_secs_f64());

    // Print collection info
    let paged = store.list_collections(0, 10, &[]).expect("Failed to list collections");
    let collections = &paged.results;
    assert!(!collections.is_empty(), "No collections found in store");
    let collection_digest = &collections[0].digest;
    eprintln!(
        "Collection: {} ({} sequences)",
        collection_digest, collections[0].n_sequences
    );

    // Step 2: Stream VRS IDs from VCF directly to stdout
    eprintln!("\nProcessing VCF: {}", vcf_path);
    println!("chrom\tpos\tref\talt\tvrs_id");
    let t1 = Instant::now();
    let count = compute_vrs_ids_streaming(&mut store, collection_digest, vcf_path, |r| {
        println!(
            "{}\t{}\t{}\t{}\t{}",
            r.chrom, r.pos, r.ref_allele, r.alt_allele, r.vrs_id
        );
    })
    .expect("Failed to compute VRS IDs");
    let vrs_time = t1.elapsed();

    eprintln!(
        "\nComputed {} VRS IDs in {:.3}s ({:.0} variants/sec)",
        count,
        vrs_time.as_secs_f64(),
        count as f64 / vrs_time.as_secs_f64()
    );
    eprintln!("Total time: {:.1}s", t0.elapsed().as_secs_f64());
}
